### ISSUE TITLE
Approve billing plan does not work

### ISSUE THREAD

**Issue by corvus** [paypal/PayPal-PHP-SDK#300]
Hi,

i have created a billing plan (patched to active state) and now i try to create a billing agreement with the following code

```php
$planID = 'P-6C301341BR608405KDOTP3IY';

$plan = new \PayPal\Api\Plan();
$plan->setId($planID);

$payer = new \PayPal\Api\Payer();
$payer->setPaymentMethod('paypal');

$agreement = new \PayPal\Api\Agreement();
$agreement->setName('Test Agreement')
    ->setDescription('Test Agreement Description')
    ->setStartDate('2015-05-05T00:00:00Z')
    ->setPlan($plan)
    ->setPayer($payer);
// with or without ->setShippingAddress($shippingAddress) makes no difference

try {
    // Please note that as the agreement has not yet activated, we wont be receiving the ID just yet.
    $agreement = $agreement->create($paypal->getApiContext());
    $approvalUrl = $agreement->getApprovalLink();
    var_dump($approvalUrl); // manually open link in new tab
} catch (Exception $e) {
    var_dump($e);
}
```

Debug log prints status code 201 and i get the approval link (https://www.sandbox.paypal.com/cgi-bin/webscr?cmd=_express-checkout&token=EC-XXXXXX) which i open in a new tab.
Now after login with a test-account (>9000€ balance, verified) i always get an error "At this time, we are unable to process your request. Please return to facilitator account's Test Store and try another option.". With other test-accounts the same error is shown.

I read in another issue that billing plans should work fine in the sandbox and on our production server with production credentials and my personal paypal account it also does not work.

I now checked multiple times with the documentation where my problem could be, but i find no error in my code.

Any ideas?!

**Comment by jaypatel512** [comment]
Hey @corvus ! 

First thing, can you please make sure you have "Billing Agreements" enabled in your developer account of which you are using the credentials.

![k8axvdfdnv](https://cloud.githubusercontent.com/assets/1309895/7455562/0caa627a-f243-11e4-8999-ea60e4790407.gif)

Once this is done, I would highly recommend trying out the similar code in samples. You could do that in your local machine by following just one simple instruction, [mentioned here](https://github.com/paypal/PayPal-PHP-SDK/wiki/Samples). 

1. Once done, try and run [Create Billing Agreement with PayPal](http://paypal.github.io/PayPal-PHP-SDK/sample/#6de667ee2194454f3298064cbf2a2290) in the sample, and see if you could follow that. 
2. If it goes through, it means, that the sample code works fine, however it would be under the default clientId and secret. 
3. Go to [samples/bootstrap.php](https://github.com/paypal/PayPal-PHP-SDK/blob/master/sample/bootstrap.php#L29-30) and replace the clientId and secret provided here with yours, and see if you are able to create that. 

Let me know if you find any trouble following these steps, and I should be able to help you resolve those.

**Comment by corvus** [comment]
Hey @jaypatel512 

"Billing Agreements" for the apps (production/test & live/sandbox) are activated.
I have tested with the sample you mentioned and get the same results.
PayPal sends back the approval url, i open it, but after the login with a test-account i get the error.
Same happens if i change clientId and secret to my own values.

Here is a screenshot from the test.
![image](https://cloud.githubusercontent.com/assets/134185/7472099/17d04056-f32f-11e4-9863-7ff69213bae5.png)

Opening the approval url works, login also but right after login i get the error...
![image](https://cloud.githubusercontent.com/assets/134185/7472121/576ed330-f32f-11e4-891a-ce8cab26b20e.png)

**Comment by jaypatel512** [comment]
Can you please see if you could pass on that Exclamation text, probably written in German. Apologies, but I am not able to recognize that language. If you could just copy paste the text, and send me, it might be helpful.

**Comment by corvus** [comment]
yes its german and means the same as "At this time, we are unable to process...."

"Wir können Ihre Anfrage zurzeit leider nicht verarbeiten. Kehren Sie zu facilitator account's Test Store zurück, und wählen Sie eine andere Option aus."

This is the debug-id header sent from paypal while creating the billing agreement, if it helps:
Paypal-Debug-Id: d521c6b01ce3c

**Comment by jaypatel512** [comment]
Awesome. We have forwarded this to API team, as it feels like you are doing all right things, and SDK is too. We will let you know as soon as I hear back from the subscription team.

**Comment by corvus** [comment]
Hey @jaypatel512 

anything new on this topic?

**Comment by jaypatel512** [comment]
Hey @corvus !

They are investigating the issue. They should have an answer really soon.

**Comment by corvus** [comment]
Hey @jaypatel512 ,

any updates?
2 weeks and paypal has still no answer oder anything?

**Comment by kangsy** [comment]
Same here. And this might be country related, because I am also from Germany.

**Comment by harshalpatel** [comment]
Not sure if this helps, but for US based account, I do not experience this issue. 

More evidence it's country related.

**Comment by jaypatel512** [comment]
Ticket has been closed, as non-reproducable in PPTIPS-3743. Please reach out to [Merchant Technical Support](https://www.paypal-techsupport.com/app/ask) directly to get more details.

**Comment by olleharstedt** [comment]
I am having this problem right now, in Germany too. Any news? Or is it simply impossible to use billing agreements in Germany?

**Comment by burki** [comment]
I am having this problem as well, also from Germany (with https://packagist.org/packages/paypal/rest-api-sdk-php#v1.5.1)

**Comment by burki** [comment]
PS: I found the following link https://developer.paypal.com/webapps/developer/docs/classic/express-checkout/integration-guide/ECRecurringPayments/ which states:

> Recurring payments cannot be created for buyers in !!!Germany or China!!. In this case, you can use reference transactions as an alternate solution.

**Comment by olleharstedt** [comment]
@burki That's for the classic API. This issue is about the REST API. Recurring payments `!=` billing plans (I think).

**Comment by burki** [comment]
@olleharstedt I just checked - if I create a test-account with country Switzerland (instead of Germany), then it works over the REST-API. So it seems to be the same country restriction.

**Comment by olleharstedt** [comment]
@burki But only for sandbox?? Why would it be closed as "non-reproducable" then?

**Comment by Kimmax** [comment]
7 Months later and this is still not fixed. Do billing agreements in Germany work in live environment? I'm using the REST Api. Not happy if I have to rewrite my whole code to use Reference Transactions..

**Comment by ppmtscory** [comment]
This is a country limitation and not a REST API limitation.  Even with the Classic API, recurring payments are not available for German consumers to set up.  This has to do with various rules and regulations that we have to follow for processing these payment types in specific countries.  The link that @burki provided in their comment above is still accurate.

**Comment by playwolf719** [comment]
same issue, is there any solution

**Comment by marcj** [comment]
This error message from Paypal is just pathetic. We as developer and user of Paypal's service are not even able to debug this as we get an extremely general error message "At this time, we are unable to process" which is not helpful at all, nor do we have a log in the developer site to dig deeper into the issue and find the root cause. Good to know at least after search for hours for a solution that I have to rewrite my code for German users.

**Comment by braebot** [comment]
@marcj, totally agree.  I'm sorry that you had this experience.  I created internal issue `PPPLMER-25138` to fix the messaging for non-supported countries.  We'll also see if we can fix the developer dashboard to restrict the countries where BAs are allowed in the first place.
